apm/monitor_test.go

```diff
@@ -0,0 +1,241 @@
+				CommandName:  "find",
+				RequestID:    42,
+				Command: buildCommand(t, birch.DC.Elements(
+					birch.EC.String("find", "jobs"),
+				)),
+			})
+		})
+		t.Run("StartEventGetMore", func(t *testing.T) {
+			collector.Started(ctx, &event.CommandStartedEvent{
+				DatabaseName: "amboy",
+				CommandName:  "getMore",
+				RequestID:    44,
+				Command: buildCommand(t, birch.DC.Elements(
+					birch.EC.String("getMore", ""),
+					birch.EC.String("collection", "jobs"),
+				)),
+			})
+			assert.Equal(t, "jobs", m.inProg[44].collName)
+		})
+		t.Run("InvalidCommand", func(t *testing.T) {
+			collector.Started(ctx, &event.CommandStartedEvent{
+				DatabaseName: "amboy",
+				CommandName:  "wat",
+				RequestID:    84,
+				Command:      nil,
+			})
+			_, ok := m.inProg[44]
+			assert.True(t, ok)
+			assert.Equal(t, "", m.inProg[84].collName)
+			assert.Equal(t, "wat", m.inProg[84].cmdName)
+		})
+		t.Run("CompleteNils", func(t *testing.T) {
+			collector.Succeeded(ctx, &event.CommandSucceededEvent{CommandFinishedEvent: event.CommandFinishedEvent{RequestID: 100}})
+			assert.Len(t, m.current, 0)
+			collector.Failed(ctx, &event.CommandFailedEvent{CommandFinishedEvent: event.CommandFinishedEvent{RequestID: 100}})
+			assert.Len(t, m.current, 0)
+		})
+		t.Run("Success", func(t *testing.T) {
+			resetMonitor(t, m)
+			collector.Started(ctx, &event.CommandStartedEvent{
+				DatabaseName: "amboy",
+				CommandName:  "find",
+				RequestID:    42,
+				Command: buildCommand(t, birch.DC.Elements(
+					birch.EC.String("find", "jobs"),
+				)),
+			})
+
+			collector.Succeeded(ctx, &event.CommandSucceededEvent{CommandFinishedEvent: event.CommandFinishedEvent{RequestID: 42}})
+			assert.Len(t, m.current, 1)
+
+			op, ok := m.current[eventKey{dbName: "amboy", cmdName: "find", collName: "jobs"}]
+			require.True(t, ok)
+			assert.EqualValues(t, 1, op.Succeeded)
+			assert.EqualValues(t, 0, op.Failed)
+		})
+		t.Run("Failed", func(t *testing.T) {
+			resetMonitor(t, m)
+			collector.Started(ctx, &event.CommandStartedEvent{
+				DatabaseName: "amboy",
+				CommandName:  "aggregate",
+				RequestID:    100,
+				Command: buildCommand(t, birch.DC.Elements(
+					birch.EC.String("aggregate", "group.jobs"),
+				)),
+			})
+
+			collector.Failed(ctx, &event.CommandFailedEvent{CommandFinishedEvent: event.CommandFinishedEvent{RequestID: 100}})
+			assert.Len(t, m.current, 1)
+
+			op, ok := m.current[eventKey{dbName: "amboy", cmdName: "aggregate", collName: "group.jobs"}]
+			require.True(t, ok)
+			assert.EqualValues(t, 0, op.Succeeded)
+			assert.EqualValues(t, 1, op.Failed)
+		})
+		t.Run("Wrapper", func(t *testing.T) {
+			t.Run("Logging", func(t *testing.T) {
+				resetMonitor(t, m)
+				nctx, ncancel := context.WithCancel(ctx)
+				defer ncancel()
+				wrapped := NewLoggingMonitor(nctx, 10*time.Millisecond, m)
+				assert.NotNil(t, wrapped)
+				assert.Implements(t, (*Monitor)(nil), wrapped)
+				time.Sleep(100 * time.Millisecond)
```

Comment:
Shouldn't we check the logs?

Reply:
this is really hard, but you're theoretically right. I'm inclined to push without it.